heartbeat/monitors/wrappers/summarizer/summarizer_test.go

```diff
@@ -219,3 +221,181 @@ func TestSummarizer(t *testing.T) {
 		})
 	}
 }
+
+// Test wrapper plugin hook order. Guaranteed order for plugins to be called upon determines
+// what data can be appended to the event at each stage through retries. With this guarantee,
+// plugins just need to ascertain that their invariants apply through hook execution order
+func TestSummarizerPluginOrder(t *testing.T) {
+	t.Parallel()
+
+	// these tests use strings to describe sequences of events
+	tests := []struct {
+		name        string
+		maxAttempts int
+		// the attempt number of the given event
+		expectedOrder []string
+	}{
+		{
+			"one attempt",
+			1,
+			[]string{"bee", "job", "ee", "bs"},
+		},
+		{
+			"two attempts",
+			2,
+			[]string{"bee", "job", "ee", "bs", "br", "bee", "job", "ee", "bs"},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Monitor setup
+			tracker := monitorstate.NewTracker(monitorstate.NilStateLoader, false)
+			sf := stdfields.StdMonitorFields{ID: "testmon", Name: "testmon", Type: "http", MaxAttempts: uint16(tt.maxAttempts)}
+
+			// Test locals
+			calls := make(chan string, 100)
+			mtx := sync.Mutex{}
+			appendCall := func(event string) {
+				mtx.Lock()
+				defer mtx.Unlock()
+
+				// Append call to global chan
+				calls <- event
+			}
+
+			// We simplify these to always down since hook order should not be
+			// determined by status
+			job := func(event *beat.Event) (j []jobs.Job, retErr error) {
+
+				calls <- "job"
+
+				event.Fields = mapstr.M{
+					"monitor": mapstr.M{
+						"id":     "test",
+						"status": string(monitorstate.StatusDown),
+					},
+				}
+
+				return nil, fmt.Errorf("dummyerr")
+			}
+
+			s := NewSummarizer(job, sf, tracker)
+			// Shorten retry delay to make tests run faster
+			s.retryDelay = 2 * time.Millisecond
+			// Add mock plugin
+			s.plugins = append(s.plugins, &MockPlugin{
+				eachEvent: func(_ *beat.Event, _ error) {
+					appendCall("ee")
+				},
+				beforeSummary: func(_ *beat.Event) {
+					appendCall("bs")
+				},
+				beforeRetry: func() {
+					appendCall("br")
+				},
+				beforeEachEvent: func(_ *beat.Event) {
+					appendCall("bee")
+				},
+			})
+			wrapped := s.Wrap(job)
+
+			_, _ = jobs.ExecJobAndConts(t, wrapped)
+
+			close(calls)
+
+			// gather order
+			rcvdOrder := []string{}
+			for c := range calls {
+				rcvdOrder = append(rcvdOrder, c)
+			}
+
+			require.Equal(t, tt.expectedOrder, rcvdOrder)
+			require.Len(t, rcvdOrder, len(tt.expectedOrder))
+		})
+	}
+}
+
+func TestRetryLightweightMonitorDuration(t *testing.T) {
+	t.Parallel()
+
+	// Monitor setup
+	tracker := monitorstate.NewTracker(monitorstate.NilStateLoader, false)
+	sf := stdfields.StdMonitorFields{ID: "testmon", Name: "testmon", Type: "http", MaxAttempts: uint16(2)}
+
+	// We simplify these to always down
+	job := func(event *beat.Event) (j []jobs.Job, retErr error) {
+		time.Sleep(1 * time.Second)
```

Comment:
nit: Is this required? Wouldnt retry delay be enough to justify the duration.

Reply:
I will say, thinking about it, it might be good to make this a constant and use that value in the test.

Reply:
I think we can just move this job a func and reuse it on both LW and Browser level tests. Not a huge deal though, if we touch this code in this PR, would make sense doing that as well.

Reply:
Hmm not sure that refactoring would change much, there're three `job := func(event *beat.Event)...` definitions on this file and they all do _slightly_ different things. 
In this case, adding `time.Sleep(1 * time.Second)` is just a way of making sure the job execution is not instantaneous is way that could trip the comparison below. It's probably not needed anyway.

This test is specifically about making sure that retry delay or previous attempts' duration is not computed on a single attempt duration, atm they are included. 👍 to adding a detailed description on what this does.